src/System.Net.Security/src/System/Net/_SecureChannel.cs

```diff
@@ -0,0 +1,1487 @@
+                if ((errorCode == Interop.SecurityStatus.OK && decspc[i].type == SecurityBufferType.Data)
+                    // or we failed to decode the data, here is the encoded data.
+                    || (errorCode != Interop.SecurityStatus.OK && decspc[i].type == SecurityBufferType.Extra))
+                {
+                    offset = decspc[i].offset;
+                    count = decspc[i].size;
+                    break;
+                }
+            }
+
+            return errorCode;
+        }
+
+        /*++
+            VerifyRemoteCertificate - Validates the content of a Remote Certificate
+
+            checkCRL if true, checks the certificate revocation list for validity.
+            checkCertName, if true checks the CN field of the certificate
+        --*/
+
+        //This method validates a remote certificate.
+        //SECURITY: The scenario is allowed in semitrust StorePermission is asserted for Chain.Build
+        //          A user callback has unique signature so it is safe to call it under permission assert.
+        //
+        internal bool VerifyRemoteCertificate(RemoteCertValidationCallback remoteCertValidationCallback)
+        {
+            GlobalLog.Enter("SecureChannel#" + Logging.HashString(this) + "::VerifyRemoteCertificate");
+            SslPolicyErrors sslPolicyErrors = SslPolicyErrors.None;
+
+            // We don't catch exceptions in this method, so it's safe for "accepted" be initialized with true.
+            bool success = false;
+            X509Chain chain = null;
+            X509Certificate2 remoteCertificateEx = null;
+
+            try
+            {
+                X509Certificate2Collection remoteCertificateStore;
+                remoteCertificateEx = GetRemoteCertificate(out remoteCertificateStore);
+                _isRemoteCertificateAvailable = remoteCertificateEx != null;
+
+                if (remoteCertificateEx == null)
+                {
+                    GlobalLog.Leave("SecureChannel#" + Logging.HashString(this) + "::VerifyRemoteCertificate (no remote cert)", (!_remoteCertRequired).ToString());
+                    sslPolicyErrors |= SslPolicyErrors.RemoteCertificateNotAvailable;
+                }
+                else
+                {
+                    chain = new X509Chain();
+                    chain.ChainPolicy.RevocationMode = _checkCertRevocation ? X509RevocationMode.Online : X509RevocationMode.NoCheck;
+                    chain.ChainPolicy.RevocationFlag = X509RevocationFlag.ExcludeRoot;
+                    if (remoteCertificateStore != null)
+                        chain.ChainPolicy.ExtraStore.AddRange(remoteCertificateStore);
+
+                    if (!chain.Build(remoteCertificateEx)       // Build failed on handle or on policy.
+                        && chain.SafeHandle.DangerousGetHandle() == IntPtr.Zero)   // Build failed to generate a valid handle.
```

Comment:
As decided, since this is the certificate handling section, it should move to a separate class which will have common/variations for the various OSs.